### ISSUE TITLE
[Bug] `attribute 'lib' missing` when attempting to use stylix

### ISSUE THREAD

**Issue by theHedgehog0** [danth/stylix#47]
I'm using stylix in a nix config, and whenever I try to evaluate my system configuration, it errors out.

The error is:
```
       error: attribute 'lib' missing

       at /nix/store/lbgfhradyynfljgmv9bywhg9h3p25qyf-modules/chromium/nixos.nix:5:5:

            4|   options.stylix.targets.chromium.enable =
            5|     config.lib.stylix.mkEnableTarget "Chromium, Google Chrome and Brave" true;
             |     ^
            6|
```

I don't know what is causing this problem, but if you can help, that would be much appreciated. Please let me know if there are any other details you need.

**Comment by theHedgehog0** [comment]
Yes I am. I will push my code changes to my forge and send a link to the relevant files.

**Comment by theHedgehog0** [comment]
Ok, so 
* The repo is [here](https://git.pyrox.dev/pyrox/nix)
* The relevant flake.nix lines are [here](https://git.pyrox.dev/pyrox/nix/src/branch/main/flake.nix#L157-L164). These pull in the stylix module from inputs and call the [configuration file for the host](https://git.pyrox.dev/pyrox/nix/src/branch/main/hosts/zaphod/configuration.nix)
* The [stylix config](https://git.pyrox.dev/pyrox/nix/src/branch/main/hosts/zaphod/theme.nix) for the host is imported on line 24 of the `configuration.nix` file listed above. All options are taken directly from the documentation. I hope this helps, please let me know if there is other relevant documentation I can provide.

**Comment by danth** [comment]
This has been reported before (#20), although it was never clear what caused it.

I would assume something outside of Stylix is the problem? You have a lot of flake inputs, and the latest commit works fine with a minimal config.

If this is the case, perhaps moving our utils into the `config.stylix` namespace rather than `config.lib` would avoid interference? Or injecting them into the real `lib`, if that is possible. However, these would both be breaking changes.

**Comment by theHedgehog0** [comment]
Well, I don't know what exactly the problem would be, but I would be glad to test any potential changes. I may fork and try to put everything in a `config.stylix` namespace and see if that works.

**Comment by dwarfmaster** [comment]
Maybe it's due to the fact that `config` is not computed yet when the options are evaluated (it should be fully lazy and work, but I admit not liking depending on things in config for the structure of options).

**Comment by jboyens** [comment]
So, I ran into a similar problem and I've got some debugging information that may help.

The error I received was:
```
       error: attribute 'xserver' missing

       at /nix/store/23gl8kb384mgqm0ys6bvcbhdpknfjlrm-modules/gnome/nixos.nix:16:17:

           15|       default = config.stylix.autoEnable
           16|              && config.services.xserver.desktopManager.gnome.enable;
             |                 ^
           17|     };
```

Now, if I--before I even load `stylix`--set `stylix.autoEnable = false` I no longer get the error. So I removed the disabling of `autoEnable` and started digging a bit. Dropping into the debugger at that point showed me that the nixos modules had yet to load and populate config with their settings/defaults. There were only _other_ pieces of my config in there. I don't have GNOME enabled (nor xserver) so they weren't populated by my config either.

The `default` of the option for the GNOME theming _feels_ suspect. I don't have any specifics or extensive experience writing nixModules to back that statement up, but using `config` in the `options` hash might be tempting the recursive gods to smite you. :stuck_out_tongue_closed_eyes: 

If I'm not mistaken, that `default` needs to evaluate as the module loads which is why the `autoEnable = false` short-circuited the check. If the GNOME / xserver bits load _first_ or the user sets those values themselves, then there's no problem because it has a value to read (either constant `false` or the user's preference) rather than an evaluated default.

In an effort to prove this out, I walked `stylix` back to ba324393e9414f57a2d0ba5a1edee5cd02b8af18 and reevaluated the flake. I ended up with this error: 

```
       error: attribute 'mkEnableTarget' missing

       at /nix/store/prq3jjd62gxlb87nlnppnjc9rxmi8rxy-modules/gnome/nixos.nix:7:5:

            6|   options.stylix.targets.gnome.enable =
            7|     config.lib.stylix.mkEnableTarget "GNOME" config.services.xserver.desktopManager.gnome.enable;
             |     ^
            8|
```

Different, yes, but still in the GNOME module. I went one further step and walked it back to ce781e1d963311216b2837e014f522c3620f0639.

**The evaluation succeeded.**

My bone-headed suggestion would be to set a default of `true`/`false` everywhere and use `lib.mkIf` around the actual changes to evaluate _later_ in the cycle to avoid the issue. Something like: `lib.mkIf (cfg.autoEnable || cfg.gnome.enable)` and see if that doesn't fix the issue. I'm going to fork `stylix` shortly to try it out for myself and see if it works.

**Comment by jboyens** [comment]
Nope. Didn't work. I do not understand the magic. Hopefully meaning can be found in the above...

**Comment by jboyens** [comment]
Couldn't leave it alone. This seems to be work great:

```
  options.stylix.targets.gnome.enable =
    lib.mkOption {
      description = lib.mdDoc "Whether to style GNOME";
      type = lib.types.bool;
      default = config.stylix.autoEnable 
             && (config.services.xserver.desktopManager.gnome.enable or false);
    };
```

Gleaned from here: https://nixos.org/manual/nix/unstable/language/operators.html#attribute-selection

**Comment by danth** [comment]
Is this still an issue?

**Comment by jboyens** [comment]
Sort of, new problems have arisen that hit before this problem so I'm not 100% sure if it's gone. 

The defaults in a lot of places are breaking because recursion is trying to use the default before the definition the default is based upon exists.

The latest instance I get (w/ commit 4da2d793e586f3f45a54fb9755ee9bf39d3cd52e)  is:
```

       error: attribute 'programs' missing

       at /nix/store/hqjk2q12m84v1aqyr9df7i34w99rraag-modules/nixvim/nixvim.nix:17:50:

           16|   config = lib.mkIf ((config.programs ? nixvim) && config.stylix.targets.nixvim.enable) (
           17|     lib.optionalAttrs (builtins.hasAttr "nixvim" options.programs) {
```
This problem can be worked around by shimming in a module prior to stylix that just defines an empty `options.programs = {}` which is enough to make the defaults get passed so that loading can continue. It will eventually evaluate correctly so the fix just tries to break an implicit load-order dependency.

Why commit 4da2d793e586f3f45a54fb9755ee9bf39d3cd52e you ask? Well, the commit AFTER that creates a separate problem:
```
       error: attribute 'mkEnableTarget' missing

       at /nix/store/pwivl3092cg504g6m06y3mns0h5343v2-modules/nixos-icons/nixos.nix:7:5:

            6|   options.stylix.targets.nixos-icons.enable =
            7|     config.lib.stylix.mkEnableTarget "the NixOS logo" true;
             |     ^
            8|
```

This one doesn't have a simple work around.

**Comment by heraldofsolace** [comment]
@jboyens I'm also running into the same issue with `mkEnableTarget` missing. From my testing, modules that set `conig.nixpkgs.overlay` are causing the breakage. These are the `nixos-icons` and `gnome` modules.

**Comment by LemmusLemmus** [comment]
I have also encountered this issue. In my case, simply adding `documentation.nixos.includeAllModules = true;` to the configuration results in this error. Not that I particularly care about enabling that option, but I guess it might be of interest.

**Comment by lane-core** [comment]
> [@jboyens](https://github.com/jboyens) I'm also running into the same issue with `mkEnableTarget` missing. From my testing, modules that set `config.nixpkgs.overlays` are causing the breakage. These are the `nixos-icons` and `gnome` modules.

New user, after removing every usage of overlays in my personal modules this error still happens. I guess I'm going to turn autoEnable off for now.

**Comment by danth** [comment]
I've mentioned this elsewhere and for other reasons: but this could be alleviated by migrating our values under `lib` into explicitly defined, read-only options

**Comment by danth** [comment]
The above would also be good for avoiding https://github.com/nix-community/nix-on-droid/issues/436

**Comment by mmlb** [comment]
I just hit this when trying out stylix, I suspect it has to do with using https://github.com/numtide/blueprint (similar to pyrox0 using snowfall?)

**Comment by awwpotato** [comment]
> I'm not a user of stylix, but this was useful context when I encountered a similar issue on another project with `documentation.nixos.includeAllModules` enabled ([StevenBlack/hosts#2828](https://github.com/StevenBlack/hosts/issues/2828)).
> 
> I found that switching from `mkEnableOption` to `mkOption` resolved the issue (https://github.com/StevenBlack/hosts/pull/2847/files), perhaps switching the implementation of `mkEnableTarget` to `mkOption` would resolve the issue here?

I made a fork with this change, could someone suffering from this issue please try switch there stylix input to `github:awwpotato/stylix/enable-mkEnable` and see if that fixes it?

**Comment by mmlb** [comment]
@awwpotato nope doesn't work for me:
```
┃        error: attribute 'lib' missing
┃        at /nix/store/d1bg28k65hhwqbng0a5gj4959s629b1n-source/stylix/overlays.nix:9:36:
┃             8| {
┃             9|   options.stylix.overlays.enable = config.lib.stylix.mkEnableTarget "packages via overlays" true;
┃              |                                    ^
┃            10|
```

**Comment by arindamukawlas** [comment]
@danth it still an issue man

**Comment by mmlb** [comment]
I ended up ripping out the `config.lib.stylix.mkEnable*` functions with claude and got my system flake to work (converted to flake-parts since my last attempt). So far its sort of working with autoEnaable = false, gtk4 and swaylock persist with errors that I'll try and track down soonish. I'm pretty sure this is not what the maintainers would want to see but if it helps to get to figure out what is going on then thats nice. My changes are https://github.com/mmlb/nix-community-stylix/commit/f0e145c21dcf0c3b0b778a972440cc7fb3f2283c

**Comment by mmlb** [comment]
I'm curious if you could expand a little bit on the following @trueNAHO @danth:
https://github.com/nix-community/stylix/pull/892#pullrequestreview-3198071432
> Purely historic reasons. IIRC, there have been some discussions with @danth about migrating config.lib.stylix to a read-only option, which would simplify our implementation, better represent the intent, and potentially be https://github.com/nix-community/stylix/issues/47. Alternatively, this could be exposed as a flake attribute.

**Comment by trueNAHO** [comment]
> I ended up ripping out the `config.lib.stylix.mkEnable*` functions with claude and got my system flake to work (converted to flake-parts since my last attempt). [...] My changes are https://github.com/mmlb/nix-community-stylix/commit/f0e145c21dcf0c3b0b778a972440cc7fb3f2283c

It seems to have basically just inlined the functions, which is rather bad for maintenance.

> I'm curious if you could expand a little bit on the following [...]:
> https://github.com/nix-community/stylix/pull/892#pullrequestreview-3198071432
> > Purely historic reasons. IIRC, there have been some discussions with [danth] about migrating config.lib.stylix to a read-only option, which would simplify our implementation, better represent the intent, and potentially be https://github.com/nix-community/stylix/issues/47. Alternatively, this could be exposed as a flake attribute.

After our `mkTarget` migration, manual

```nix
enable = config.lib.stylix.mkEnableTarget /* ... */;
```

declarations are internally gone and `config.lib.stylix.mkEnableTarget` can eventually be removed. There should be a soft deprecation phase because other projects might rely on this function.

Unlike the `config.lib.stylix.mkEnableTarget` function, it might not be possible to generally remove `config.lib.stylix.mkEnableWallpaper`.

AFAIK, read-only option interfacing is currently only done for:

- https://github.com/nix-community/stylix/blob/5a5d0f865788a364f458e63f808c2f3df0d33e39/modules/neovim/nixvim.nix#L35-L42
- https://github.com/nix-community/stylix/blob/5a5d0f865788a364f458e63f808c2f3df0d33e39/modules/sway/hm.nix#L14-L31

The read-only option is better than an arbitrary attribute value because the exported function is not supposed to be modified internally or externally. The implementation is simplified because the exporting is handled by the Nixpkgs option system.

On a side note, it would be great to have a minimal (not your entire NixOS repo) reproducible example for this issue. Ideally, as a testbed to prevent the issue steming from seemingly totally unrelated things. AFAIK, CI has also never suffered from this issue.

**Comment by mmlb** [comment]
> > I ended up ripping out the `config.lib.stylix.mkEnable*` functions with claude and got my system flake to work (converted to flake-parts since my last attempt). [...] My changes are [mmlb@f0e145c](https://github.com/mmlb/nix-community-stylix/commit/f0e145c21dcf0c3b0b778a972440cc7fb3f2283c)
> 
> It seems to have basically just inlined the functions, which is rather bad for maintenance.

Agree, was just posting that as a sort of unblocking for triaging. 

> On a side note, it would be great to have a minimal (not your entire NixOS repo) reproducible example for this issue. Ideally, as a testbed to prevent the issue steming from seemingly totally unrelated things. AFAIK, CI has also never suffered from this issue.

Now that I have something that sort-of works can make progress on a minimal reproducer. I'll see about adding a flake-parts reproducer to use with the testbed setup.

**Comment by mmlb** [comment]
So I just tried rebasing my fork and the diff was too much. I decided to let claude try to do the reproducer and it honed in on what seems to be the trigger. Looks like the culprit is not flake-parts nor blueprint but `documentation.nixos.includeAllModules = true`! I went and checked @pyrox0's repo from around that time and came upon same config: https://git.pyrox.dev/pyrox/nix/src/commit/32f57419cfdaa14d69cc45787220fc90ac678dfb/hosts/zaphod/misc.nix#L21

So here's my repro repo https://github.com/mmlb/stylix-missing-lib-repro/
